File: scripts/sbbf_k_sensitivity.py
#!/usr/bin/env python3
"""
SBBF K-Sensitivity Benchmark for Voxelized Meshes

Measures FPR as a function of k (hash functions) for SBBF on voxelized 3D meshes.
Tests k = 2..12 with constant fill rate by adjusting memory per k.

Configurations:
- Meshes: bunny_128, teapot_128, armadillo_128, dragon_128
- SFC type: HILBERT_3D (best for structured spatial data)
- Intra-block strategies: double_hash, pattern_lookup
- Seed strategies: XOR, MULTIPLY_SHIFT

Usage:
    pip install sbbf[benchmarks]
    python scripts/sbbf_k_sensitivity.py
"""

import json
import math
import sys
from datetime import datetime
from pathlib import Path

import h5py
import numpy as np

try:
    import sbbf
except ImportError as e:
    print("ERROR: Could not import sbbf module.")
    print("Install with: pip install sbbf")
    print(f"Details: {e}")
    sys.exit(1)

# ============================================================================
# Configuration
# ============================================================================

PROJECT_ROOT = Path(__file__).parent.parent
HDF5_DIR = PROJECT_ROOT / "datasets" / "hdf5"
OUTPUT_DIR = PROJECT_ROOT / "sbbf_results" / "k_sensitivity"
TARGET_FILL_RATE = 0.5

# K values to test
K_VALUES = [2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12]

# Meshes to test
MESHES = [
    {"name": "bunny", "resolution": 128},
    {"name": "teapot", "resolution": 128},
    {"name": "armadillo", "resolution": 128},
    {"name": "dragon", "resolution": 128},
]

# Configurations
INTRA_STRATEGIES = [
    ("double_hash", sbbf.IntraBlockStrategy.DOUBLE_HASH),
    ("pattern_lookup", sbbf.IntraBlockStrategy.PATTERN_LOOKUP),
]

SEED_STRATEGIES = [
    ("XOR", sbbf.SeedStrategy.XOR),
    ("MULTIPLY_SHIFT", sbbf.SeedStrategy.MULTIPLY_SHIFT),
]


# ============================================================================
# Helper Functions
# ============================================================================


def compute_log_blocks(n: int, k: int, target_fill: float = 0.5, bits_per_block: int = 64) -> int:
    """Compute log_num_blocks to achieve target fill rate."""
    required_blocks = (n * k) / (target_fill * bits_per_block)
    log_blocks = int(math.ceil(math.log2(required_blocks)))
    return max(log_blocks, 10)


def load_voxels(mesh_name: str, resolution: int) -> np.ndarray:
    """Load voxel coordinates from HDF5 file."""
    path = HDF5_DIR / f"{mesh_name}_{resolution}.h5"
    if not path.exists():
        raise FileNotFoundError(f"Voxel file not found: {path}")

    with h5py.File(path, "r") as f:
        coords = f["coords"][:]
    return coords


# ============================================================================
# Benchmark Functions
# ============================================================================


def run_benchmark(
    voxel_coords: np.ndarray,
    resolution: int,
    k: int,
    intra_strategy: sbbf.IntraBlockStrategy,
    seed_strategy: sbbf.SeedStrategy,
) -> dict:
    """Run SBBF benchmark for a single configuration."""
    num_voxels = len(voxel_coords)
    num_negatives = resolution**3 - num_voxels
    true_set = set(map(tuple, voxel_coords))

    log_num_blocks = compute_log_blocks(num_voxels, k, TARGET_FILL_RATE)

    config = sbbf.SBBFConfig()
    config.sfc_type = sbbf.SFCType.HILBERT_3D
    config.log_num_blocks = log_num_blocks
    config.hash_k = k
    config.bits_per_block = 64
    config.sfc_bits = max(8, int(math.ceil(math.log2(resolution + 1))))
    config.intra_strategy = intra_strategy
    config.seed_strategy = seed_strategy
    if intra_strategy == sbbf.IntraBlockStrategy.PATTERN_LOOKUP:
        config.pattern_table_size = 1024

    bf = sbbf.SpatialBlockedBloomFilter(config)

    # Insert all voxels
    for x, y, z in voxel_coords:
        bf.put3d(int(x), int(y), int(z))

    # Query all grid points to measure FPR
    false_positives = 0
    for x in range(resolution):
        for y in range(resolution):
            for z in range(resolution):
                if bf.query3d(x, y, z):
                    if (x, y, z) not in true_set:
                        false_positives += 1

    fpr = false_positives / num_negatives if num_negatives > 0 else 0.0

    return {
        "fpr": fpr,
        "false_positives": false_positives,
        "actual_fill_rate": bf.fill_ratio(),
        "memory_bytes": bf.memory_bytes(),
        "log_num_blocks": log_num_blocks,
    }


# ============================================================================
# Main
# ============================================================================


def main():
    print("=" * 60)
    print("SBBF K-Sensitivity Benchmark (Voxelized Meshes)")
    print("=" * 60)
    print(f"Target fill rate: {TARGET_FILL_RATE}")
    print(f"K values: {K_VALUES}")
    print(f"Meshes: {[m['name'] for m in MESHES]}")
    print("=" * 60)

    OUTPUT_DIR.mkdir(parents=True, exist_ok=True)

    results = []
    available_meshes = []

    for mesh in MESHES:
        mesh_name = mesh["name"]
        resolution = mesh["resolution"]
        path = HDF5_DIR / f"{mesh_name}_{resolution}.h5"
        if path.exists():
            available_meshes.append(mesh)
        else:
            print(f"Warning: {mesh_name}_{resolution}.h5 not found, skipping.")

    if not available_meshes:
        print("ERROR: No voxel datasets found!")
        print("Run: python scripts/download_voxel_meshes.py")
        sys.exit(1)

    total_configs = len(available_meshes) * len(K_VALUES) * len(INTRA_STRATEGIES) * len(SEED_STRATEGIES)
    current = 0

    for mesh in available_meshes:
        mesh_name = mesh["name"]
        resolution = mesh["resolution"]

        print(f"\n--- {mesh_name}_{resolution} ---")

        try:
            voxel_coords = load_voxels(mesh_name, resolution)
            print(f"Loaded {len(voxel_coords):,} voxels")
        except FileNotFoundError as e:
            print(f"SKIPPED: {e}")
            continue

        for k in K_VALUES:
            for intra_name, intra_strategy in INTRA_STRATEGIES:
                for seed_name, seed_strategy in SEED_STRATEGIES:
                    current += 1
                    print(
                        f"[{current}/{total_configs}] k={k} {intra_name} {seed_name}...",
                        end=" ",
                        flush=True,
                    )

                    result = run_benchmark(
                        voxel_coords,
                        resolution,
                        k,
                        intra_strategy,
                        seed_strategy,
                    )

                    print(f"FPR={result['fpr']:.4%}")

                    results.append(
                        {
                            "config": {
                                "mesh": mesh_name,
                                "resolution": resolution,
                                "k": k,
                                "sfc_type": "HILBERT_3D",
                                "intra_strategy": intra_name,
                                "seed_strategy": seed_name,
                                "log_num_blocks": result["log_num_blocks"],
                            },
                            "metrics": {
                                "fpr": result["fpr"],
                                "false_positives": result["false_positives"],
                                "actual_fill_rate": result["actual_fill_rate"],
                                "memory_bytes": result["memory_bytes"],
                            },
                        }
                    )

    output = {
        "experiment": "k_sensitivity_voxel",
        "timestamp": datetime.now().isoformat(),
        "config": {
            "target_fill_rate": TARGET_FILL_RATE,
            "k_values": K_VALUES,
            "meshes": available_meshes,
        },
        "results": results,
    }

    output_path = OUTPUT_DIR / "voxel.json"
    with open(output_path, "w") as f:
        json.dump(output, f, indent=2)

    print("\n" + "=" * 60)
    print(f"Results saved to: {output_path}")
    print(f"Total configurations: {len(results)}")
    print("=" * 60)


if __name__ == "__main__":
    main()

File: benchmarks/sbbf_k_sensitivity.cpp
/*
 * SBBF K-Sensitivity Benchmark
 *
 * Measures FPR as a function of k (hash functions) for SBBF variants.
 * Tests k = 2..12 with constant fill rate by adjusting memory per k.
 *
 * Configurations:
 * - SFC types: Morton2D, Morton3D, Hilbert2D, Hilbert3D
 * - Intra-block strategies: double_hash, pattern_lookup
 * - Seed strategies: XOR, MULTIPLY_SHIFT
 *
 * Output: JSON file with FPR for each configuration
 */

#include <sbbf/spatial_blocked_bloom_filter.hpp>
#include "json.hpp"

#include <algorithm>
#include <chrono>
#include <cmath>
#include <fstream>
#include <iomanip>
#include <iostream>
#include <random>
#include <sstream>
#include <sys/stat.h>
#include <vector>

using json = nlohmann::json;
using namespace sbbf;

// ============================================================================
// Configuration
// ============================================================================

const size_t NUM_ELEMENTS = 100000;
const size_t NUM_QUERIES = 100000;
const double TARGET_FILL_RATE = 0.5;
const unsigned COORD_BITS = 16;  // Must be multiple of 4 for Hilbert
const uint32_t MAX_COORD = (1 << COORD_BITS) - 1;
const std::string OUTPUT_DIR = "./sbbf_results/k_sensitivity/";

// K values to test
const std::vector<unsigned> K_VALUES = {2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12};

// ============================================================================
// Data Structures
// ============================================================================

struct Point2D {
    uint32_t x, y;
};

struct Point3D {
    uint32_t x, y, z;
};

// ============================================================================
// Helper Functions
// ============================================================================

std::string get_iso_timestamp() {
    auto now = std::chrono::system_clock::now();
    auto time = std::chrono::system_clock::to_time_t(now);
    std::tm tm = *std::gmtime(&time);
    std::ostringstream oss;
    oss << std::put_time(&tm, "%Y-%m-%dT%H:%M:%SZ");
    return oss.str();
}

void ensure_dir(const std::string& path) {
    mkdir(path.c_str(), 0755);
}

// Compute log_num_blocks to achieve target fill rate
// Fill rate = (n * k) / (num_blocks * bits_per_block)
// For target fill f: num_blocks = (n * k) / (f * bits_per_block)
unsigned compute_log_blocks(size_t n, unsigned k, double target_fill = 0.5, unsigned bits_per_block = 64) {
    double required_blocks = static_cast<double>(n * k) / (target_fill * bits_per_block);
    unsigned log_blocks = static_cast<unsigned>(std::ceil(std::log2(required_blocks)));
    return std::max(log_blocks, 10u);  // Minimum 2^10 = 1024 blocks
}

// ============================================================================
// Data Generation
// ============================================================================

std::vector<Point2D> generate_uniform_2d(size_t n, uint32_t max_coord, uint64_t seed) {
    std::mt19937_64 rng(seed);
    std::uniform_int_distribution<uint32_t> dist(0, max_coord);
    std::vector<Point2D> points;
    points.reserve(n);
    for (size_t i = 0; i < n; ++i) {
        points.push_back({dist(rng), dist(rng)});
    }
    return points;
}

std::vector<Point3D> generate_uniform_3d(size_t n, uint32_t max_coord, uint64_t seed) {
    std::mt19937_64 rng(seed);
    std::uniform_int_distribution<uint32_t> dist(0, max_coord);
    std::vector<Point3D> points;
    points.reserve(n);
    for (size_t i = 0; i < n; ++i) {
        points.push_back({dist(rng), dist(rng), dist(rng)});
    }
    return points;
}

// ============================================================================
// Benchmark Functions
// ============================================================================

struct BenchResult {
    unsigned k;
    std::string sfc_type;
    std::string intra_strategy;
    std::string seed_strategy;
    unsigned dimensions;
    unsigned log_num_blocks;
    size_t memory_bytes;
    double actual_fill_rate;
    double fpr;
};

std::string sfc_type_to_string(SFCType sfc) {
    switch (sfc) {
        case SFCType::MORTON_2D: return "MORTON_2D";
        case SFCType::MORTON_3D: return "MORTON_3D";
        case SFCType::HILBERT_2D: return "HILBERT_2D";
        case SFCType::HILBERT_3D: return "HILBERT_3D";
    }
    return "UNKNOWN";
}

std::string intra_strategy_to_string(IntraBlockStrategy strategy) {
    switch (strategy) {
        case IntraBlockStrategy::DOUBLE_HASH: return "double_hash";
        case IntraBlockStrategy::PATTERN_LOOKUP: return "pattern_lookup";
        case IntraBlockStrategy::MULTIPLEXED: return "multiplexed";
    }
    return "unknown";
}

std::string seed_strategy_to_string(SeedStrategy strategy) {
    switch (strategy) {
        case SeedStrategy::XOR: return "XOR";
        case SeedStrategy::MULTIPLY_SHIFT: return "MULTIPLY_SHIFT";
    }
    return "unknown";
}

template<unsigned SFCBits>
BenchResult run_benchmark_2d(unsigned k, SFCType sfc_type,
                              IntraBlockStrategy intra_strategy,
                              SeedStrategy seed_strategy,
                              const std::vector<Point2D>& insert_data,
                              const std::vector<Point2D>& query_data) {
    unsigned log_blocks = compute_log_blocks(insert_data.size(), k, TARGET_FILL_RATE);

    SBBFConfig conf;
    conf.sfc_type = sfc_type;
    conf.sfc_bits = SFCBits;
    conf.log_num_blocks = log_blocks;
    conf.hash_k = k;
    conf.bits_per_block = 64;
    conf.intra_strategy = intra_strategy;
    conf.seed_strategy = seed_strategy;
    if (intra_strategy == IntraBlockStrategy::PATTERN_LOOKUP) {
        conf.pattern_table_size = 1024;
    }

    SpatialBlockedBloomFilter64<SFCBits> filter(conf);

    // Insert all elements
    for (const auto& p : insert_data) {
        filter.put2D(p.x, p.y);
    }

    // Measure FPR on query data (points not in filter)
    size_t false_positives = 0;
    for (const auto& p : query_data) {
        if (filter.get_bool_2D(p.x, p.y)) ++false_positives;
    }

    BenchResult result;
    result.k = k;
    result.sfc_type = sfc_type_to_string(sfc_type);
    result.intra_strategy = intra_strategy_to_string(intra_strategy);
    result.seed_strategy = seed_strategy_to_string(seed_strategy);
    result.dimensions = 2;
    result.log_num_blocks = log_blocks;
    result.memory_bytes = filter.memory_usage();
    result.actual_fill_rate = filter.block_fill_ratio();
    result.fpr = static_cast<double>(false_positives) / query_data.size();

    return result;
}

template<unsigned SFCBits>
BenchResult run_benchmark_3d(unsigned k, SFCType sfc_type,
                              IntraBlockStrategy intra_strategy,
                              SeedStrategy seed_strategy,
                              const std::vector<Point3D>& insert_data,
                              const std::vector<Point3D>& query_data) {
    unsigned log_blocks = compute_log_blocks(insert_data.size(), k, TARGET_FILL_RATE);

    SBBFConfig conf;
    conf.sfc_type = sfc_type;
    conf.sfc_bits = SFCBits;
    conf.log_num_blocks = log_blocks;
    conf.hash_k = k;
    conf.bits_per_block = 64;
    conf.intra_strategy = intra_strategy;
    conf.seed_strategy = seed_strategy;
    if (intra_strategy == IntraBlockStrategy::PATTERN_LOOKUP) {
        conf.pattern_table_size = 1024;
    }

    SpatialBlockedBloomFilter64<SFCBits> filter(conf);

    // Insert all elements
    for (const auto& p : insert_data) {
        filter.put3D(p.x, p.y, p.z);
    }

    // Measure FPR on query data
    size_t false_positives = 0;
    for (const auto& p : query_data) {
        if (filter.get_bool_3D(p.x, p.y, p.z)) ++false_positives;
    }

    BenchResult result;
    result.k = k;
    result.sfc_type = sfc_type_to_string(sfc_type);
    result.intra_strategy = intra_strategy_to_string(intra_strategy);
    result.seed_strategy = seed_strategy_to_string(seed_strategy);
    result.dimensions = 3;
    result.log_num_blocks = log_blocks;
    result.memory_bytes = filter.memory_usage();
    result.actual_fill_rate = filter.block_fill_ratio();
    result.fpr = static_cast<double>(false_positives) / query_data.size();

    return result;
}

// ============================================================================
// Main
// ============================================================================

int main() {
    std::cout << "========================================\n";
    std::cout << "SBBF K-Sensitivity Benchmark\n";
    std::cout << "========================================\n";
    std::cout << "Elements: " << NUM_ELEMENTS << "\n";
    std::cout << "Queries: " << NUM_QUERIES << "\n";
    std::cout << "Target fill rate: " << TARGET_FILL_RATE << "\n";
    std::cout << "Coord bits: " << COORD_BITS << "\n";
    std::cout << "K values: ";
    for (auto k : K_VALUES) std::cout << k << " ";
    std::cout << "\n";
    std::cout << "========================================\n\n";

    ensure_dir("./sbbf_results");
    ensure_dir(OUTPUT_DIR);

    // Generate data
    std::cout << "Generating synthetic data...\n";
    auto insert_2d = generate_uniform_2d(NUM_ELEMENTS, MAX_COORD, 42);
    auto query_2d = generate_uniform_2d(NUM_QUERIES, MAX_COORD, 123);  // Different seed!
    auto insert_3d = generate_uniform_3d(NUM_ELEMENTS, MAX_COORD, 42);
    auto query_3d = generate_uniform_3d(NUM_QUERIES, MAX_COORD, 123);
    std::cout << "Done.\n\n";

    std::vector<BenchResult> results;

    // Configuration matrix
    std::vector<std::pair<SFCType, unsigned>> sfc_configs = {
        {SFCType::MORTON_2D, 2},
        {SFCType::HILBERT_2D, 2},
        {SFCType::MORTON_3D, 3},
        {SFCType::HILBERT_3D, 3}
    };
    std::vector<IntraBlockStrategy> intra_strategies = {
        IntraBlockStrategy::DOUBLE_HASH,
        IntraBlockStrategy::PATTERN_LOOKUP
    };
    std::vector<SeedStrategy> seed_strategies = {
        SeedStrategy::XOR,
        SeedStrategy::MULTIPLY_SHIFT
    };

    size_t total_configs = K_VALUES.size() * sfc_configs.size() *
                           intra_strategies.size() * seed_strategies.size();
    size_t current = 0;

    // Run all configurations
    for (auto k : K_VALUES) {
        for (const auto& [sfc_type, dim] : sfc_configs) {
            for (auto intra : intra_strategies) {
                for (auto seed : seed_strategies) {
                    ++current;
                    std::cout << "[" << current << "/" << total_configs << "] "
                              << "k=" << k
                              << " " << sfc_type_to_string(sfc_type)
                              << " " << intra_strategy_to_string(intra)
                              << " " << seed_strategy_to_string(seed)
                              << "... ";
                    std::cout.flush();

                    BenchResult result;
                    if (dim == 2) {
                        result = run_benchmark_2d<COORD_BITS>(
                            k, sfc_type, intra, seed, insert_2d, query_2d);
                    } else {
                        result = run_benchmark_3d<COORD_BITS>(
                            k, sfc_type, intra, seed, insert_3d, query_3d);
                    }

                    std::cout << "FPR=" << std::fixed << std::setprecision(4)
                              << (result.fpr * 100) << "%\n";

                    results.push_back(result);
                }
            }
        }
    }

    // Save results to JSON
    json output;
    output["experiment"] = "k_sensitivity";
    output["timestamp"] = get_iso_timestamp();
    output["config"] = {
        {"num_elements", NUM_ELEMENTS},
        {"num_queries", NUM_QUERIES},
        {"target_fill_rate", TARGET_FILL_RATE},
        {"coord_bits", COORD_BITS},
        {"k_values", K_VALUES}
    };

    json results_json = json::array();
    for (const auto& r : results) {
        results_json.push_back({
            {"config", {
                {"k", r.k},
                {"sfc_type", r.sfc_type},
                {"intra_strategy", r.intra_strategy},
                {"seed_strategy", r.seed_strategy},
                {"dimensions", r.dimensions},
                {"log_num_blocks", r.log_num_blocks}
            }},
            {"metrics", {
                {"fpr", r.fpr},
                {"actual_fill_rate", r.actual_fill_rate},
                {"memory_bytes", r.memory_bytes}
            }}
        });
    }
    output["results"] = results_json;

    std::string output_path = OUTPUT_DIR + "synthetic.json";
    std::ofstream out(output_path);
    out << output.dump(2);
    out.close();

    std::cout << "\n========================================\n";
    std::cout << "Results saved to: " << output_path << "\n";
    std::cout << "Total configurations: " << results.size() << "\n";
    std::cout << "========================================\n";

    return 0;
}
